Source/CarthageKit/DiagnosticResolver.swift

```diff
@@ -0,0 +1,197 @@
+    private let versionsForDependency: (Dependency) -> SignalProducer<PinnedVersion, CarthageError>
+    private let resolvedGitReference: (Dependency, String) -> SignalProducer<PinnedVersion, CarthageError>
+    private let dependenciesForDependency: (Dependency, PinnedVersion) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError>
+
+    private var versionSets = [Dependency: [PinnedVersion]]()
+    private var handledDependencies = Set<PinnedDependency>()
+
+    public var ignoreErrors: Bool = false
+
+    // Specify mappings to anonimize private dependencies (which may not be disclosed as part of the diagnostics)
+    public var dependencyMappings: [Dependency: Dependency]?
+
+    public let diagnosticResolverEvents: Signal<DiagnosticResolverEvent, NoError>
+    private let diagnosticResolverEventPublisher: Signal<DiagnosticResolverEvent, NoError>.Observer
+
+    private enum DiagnosticResolverError: Error {
+        case versionRetrievalFailure(message: String)
+        case dependencyRetrievalFailure(message: String)
+    }
+
+    public convenience init(versionsForDependency: @escaping (Dependency) -> SignalProducer<PinnedVersion, CarthageError>, dependenciesForDependency: @escaping (Dependency, PinnedVersion) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError>, resolvedGitReference: @escaping (Dependency, String) -> SignalProducer<PinnedVersion, CarthageError>) {
+        self.init(versionsForDependency: versionsForDependency,
+                  dependenciesForDependency: dependenciesForDependency,
+                  resolvedGitReference: resolvedGitReference,
+                  localRepository: LocalRepository(directoryURL: URL(fileURLWithPath: "/tmp")))
+    }
+
+    public init(
+        versionsForDependency: @escaping (Dependency) -> SignalProducer<PinnedVersion, CarthageError>,
+        dependenciesForDependency: @escaping (Dependency, PinnedVersion) -> SignalProducer<(Dependency, VersionSpecifier), CarthageError>,
+        resolvedGitReference: @escaping (Dependency, String) -> SignalProducer<PinnedVersion, CarthageError>,
+        localRepository: LocalRepository) {
+        self.versionsForDependency = versionsForDependency
+        self.dependenciesForDependency = dependenciesForDependency
+        self.resolvedGitReference = resolvedGitReference
+        self.localRepository = localRepository
+
+        let (signal, observer) = Signal<DiagnosticResolverEvent, NoError>.pipe()
+        diagnosticResolverEvents = signal
+        diagnosticResolverEventPublisher = observer
+    }
+
+    public func resolve(
+        dependencies: [Dependency: VersionSpecifier],
+        lastResolved: [Dependency: PinnedVersion]? = nil,
+        dependenciesToUpdate: [String]? = nil
+        ) -> SignalProducer<[Dependency: PinnedVersion], CarthageError> {
+        let result: Result<[Dependency: PinnedVersion], CarthageError>
+        do {
+            try traverse(dependencies: Array(dependencies))
```

Comment:
The signature of `traverse` seems a missed opportunity. It could return the successful processed dependencies while instead it's just waiting to fail or otherwise return an empty map.

Overall I'm not a fan of `throw`ing. I would rather use Result.

Reply:
I decided to let this class not implement the resolver protocol, since it is not really a resolver. Renamed it to DependencyCrawler instead. It now returns an empty result or a CarthageError. See the updated class.